core/src/main/scala/org/http4s/StaticFile.scala

```diff
@@ -66,31 +66,37 @@ object StaticFile {
 
   def fromURL[F[_]](url: URL, blocker: Blocker, req: Option[Request[F]] = None)(implicit
       F: Sync[F],
-      cs: ContextShift[F]): OptionT[F, Response[F]] =
-    OptionT.liftF(F.delay {
-      val urlConn = url.openConnection
-      val lastmod = HttpDate.fromEpochSecond(urlConn.getLastModified / 1000).toOption
-      val ifModifiedSince = req.flatMap(_.headers.get(`If-Modified-Since`))
-      val expired = (ifModifiedSince, lastmod).mapN(_.date < _).getOrElse(true)
-
-      if (expired) {
-        val lastModHeader: List[Header] = lastmod.map(`Last-Modified`(_)).toList
-        val contentType = nameToContentType(url.getPath).toList
-        val len = urlConn.getContentLengthLong
-        val lenHeader =
-          if (len >= 0) `Content-Length`.unsafeFromLong(len)
-          else `Transfer-Encoding`(TransferCoding.chunked)
-        val headers = Headers(lenHeader :: lastModHeader ::: contentType)
-
-        Response(
-          headers = headers,
-          body = readInputStream[F](F.delay(url.openStream), DefaultBufferSize, blocker)
-        )
-      } else {
-        urlConn.getInputStream.close()
-        Response(NotModified)
-      }
-    })
+      cs: ContextShift[F]): OptionT[F, Response[F]] = {
+    val fileUrl = url.getFile()
+    val file = new File(fileUrl)
+    if (file.isDirectory())
```

Comment:
This `isDirectory()` call is a side effect.  (I _think_ the two above it are not, as they don't consult the file system yet, but I may be wrong.)  What if we do the following?

* move this into the `F.delay`
* make the `F.delay` return an `Option`. This case returns `None`, and what's there returns `Some`.
* replace the `liftF` with `apply`.

Reply:
Resolved this. Verified that `val fileUrl = url.getFile()` and `val file = new File(fileUrl)` don't perform any side effects.